### PR TITLE
pandoc: Update to 2.7.3 with major Portfile refactoring

### PR DESCRIPTION
**Pull request by essandess**
pandoc: Update to 2.7.3 with major Portfile refactoring

* Use GitHub release
* Build using Haskell stack tool
* Eliminate dependencies on multiple out-of-date hs-* ports
* See https://github.com/macports/macports-ports/pull/4633

#### Description

Notes:
* This will not build until [port:stack](https://github.com/macports/macports-ports/pull/4633) is committed
* The out-of-date port dependencies `hs-*` are all deprecated and likely unnecessary
* cc @neverpanic @pmetzger @mojca 

###### Type(s)

- [x] update
- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.14.5 18F132
Xcode 10.2.1 10E1001

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by cjones051073** [comment]
Making this WIP, until https://github.com/macports/macports-ports/pull/4633 merged, just to prevent accidental merger before it is tested.

**Comment by amake** [comment]
Does this need to still have the haskell 1.0 portgroup if it's not using `haskell.setup`?

If stack is to be the new way to do everything Haskell, then I would suggest making a new haskell portgroup (1.1 or 2.0 or whatever) that handles the build phase. This would facilitate making other stack-based ports, and also allow keeping the current Haskell ports working with the old ghc (per my comment in #5050).

**Comment by essandess** [comment]
> Does this need to still have the haskell 1.0 portgroup if it's not using `haskell.setup`?
> 
> If stack is to be the new way to do everything Haskell, then I would suggest making a new haskell portgroup (1.1 or 2.0 or whatever) that handles the build phase. This would facilitate making other stack-based ports, and also allow keeping the current Haskell ports working with the old ghc (per my comment in #5050).

That’s a good suggestion. Personally, I still don’t understand what portgroups are or how to write them. And even if I did, would want to integrate some practical experience with a few `stack`-based build before writing a general recipe for them.

**Comment by pmetzger** [comment]
> That’s a good suggestion. Personally, I still don’t understand what portgroups are or how to write them.

I'd suggest reading the documentation for them, then. Portgroups are effectively libraries of tcl code used to automate common functions for groups of ports. There's nothing very hard about them; they allow factoring out of functionality needed for many separate ports. For example, the `github` portgroup handles common functions any port whose sources are hosted on github might want. As I recall, you promised to construct a portgroup for gpg signature checks when I removed my objection to the integration of a port that had gpg signature checking machinery.

**Comment by essandess** [comment]
@pmetzger  
> I'd suggest reading the documentation for them, then. Portgroups are effectively libraries of tcl code used to automate common functions for groups of ports. There's nothing very hard about them; they allow factoring out of functionality needed for many separate ports. For example, the `github` portgroup handles common functions any port whose sources are hosted on github might want. As I recall, you promised to construct a portgroup for gpg signature checks when I removed my objection to the integration of a port that had gpg signature checking machinery.

Yes, we agreed in the recent julia port upgrade, https://github.com/macports/macports-ports/pull/4825#issuecomment-519892571.

I see documentation for using some of the portgroups that @cjones051073 linked to, but not for writing them. I’ve run into these specific difficulties:

1. How/where does one write a local portgroup so that it can be tested in a local portfile? I don’t see any documentation for this or an obvious way to do it. Perhaps it’s obvious but I’m missing it.
2. How does one write (or copy) portgroup functions that does these two things:
```
verify_gpg_signature yes
```
sets a global `verify_gpg_signature_flag` to true? And defines the global proc `verify_gpg_signature`?

**Comment by pmetzger** [comment]
> I see documentation for using some of the portgroups that @cjones051073 linked to, but not for writing them.

Beyond looking at existing portgroups (and knowing some tcl, which is vital), you can ask on the macports developers mailing list for specific questions. Generally, though, the first questions you were asking were macports specific, but the last questions are basically tcl questions.

**Comment by essandess** [comment]
> Beyond looking at existing portgroups …

@pmetzger @cjones051073 I just gave it my best shot in https://github.com/macports/macports-ports/pull/5124

@pmetzger Would you please drive a decision getting `stack` PR https://github.com/macports/macports-ports/pull/4633 and `cabal` PR https://github.com/macports/macports-ports/pull/4715 merged as soon as possible? Many existing Haskell ports are broken and years out of date and need modern tools to fix them now that `ghc` has been updated. Also see https://github.com/macports/macports-ports/pull/5050.

**Comment by pmetzger** [comment]
I don't have time to drive that; it would require more attention span than I've got. I suggest trying to get someone on -developers to act to help with shepherding it through.

**Comment by essandess** [comment]
> I don't have time to drive that; it would require more attention span than I've got. I suggest trying to get someone on -developers to act to help with shepherding it through.

@pmetzger, @cjones051073 [says](https://github.com/macports/macports-ports/pull/4633#issuecomment-522343942) “@pmetzger needs to agree to the changes before this can be merged.”

Would you please at least agree to the changes so that we can move forward fixing other stuff?

**Comment by essandess** [comment]
With the merge of port `stack` this works and is no longer WIP.

@mf2k @cjones051073 @neverpanic Would you please merge this? It's ready to go.

File: textproc/pandoc/Portfile [hunk 1]
```diff
@@ -24,33 +25,45 @@ long_description    \
...
-                    port:hs-unordered-containers \
-                    port:hs-vector \
-                    port:hs-xml \
-                    port:hs-yaml \
-                    port:hs-zip-archive \
-                    port:hs-zlib
+depends_build-append  \
+                    port:stack
+
+set system_gcc      /usr/bin/gcc
```
Comment (mojca) [inline textproc/pandoc/Portfile]: Why is this needed? Why cannot other compilers be used?

File: textproc/pandoc/Portfile [hunk 2]
```diff
@@ -24,33 +25,45 @@ long_description    \
...
+# the PATH environment must provide /usr/bin/gcc, not ${prefix}/bin/gcc
+build.env \
+    PATH=/usr/bin:$env(PATH) \
+    PREFIX=${prefix} \
+    CC=${system_gcc}
+
+build {
+    xinstall -m 0755 -d ${workpath}/.stack
+    # the PATH environment must use /usr/bin/gcc, not ${prefix}/bin/gcc
+    system -W ${worksrcpath} \
```
Comment (mojca) [inline textproc/pandoc/Portfile]: This looks as if it could be split into configure, build, destroot phase? Not that it is wrong this way.

File: textproc/pandoc/Portfile [hunk 3]
```diff
@@ -24,33 +25,45 @@ long_description    \
...
+        ${destroot}${prefix}/bin
+}
+
+# the PATH environment must provide /usr/bin/gcc, not ${prefix}/bin/gcc
+test.env \
+    PATH=/usr/bin:$env(PATH) \
+    PREFIX=${prefix} \
+    CC=${system_gcc}
+
+test {
```
Comment (mojca) [inline textproc/pandoc/Portfile]: I guess you need `test.run yes` to enable tests?

File: textproc/pandoc/Portfile [hunk 4]
```diff
@@ -24,33 +25,45 @@ long_description    \
...
+    system -W ${worksrcpath} \
+        "${prefix}/bin/stack test --test-arguments='-p markdown'"
```
Comment (mojca) [inline textproc/pandoc/Portfile]: Maybe one could point `test.cmd` to stack and use other test.something variables here.

File: textproc/pandoc/Portfile (hunk 1 above)
Comment (cjones051073) [reply textproc/pandoc/Portfile]: Indeed. Please do nit hardcode a specific compiler like this. Any reason why this cannot just be ${configure.cc} ?

Comment (essandess) [reply textproc/pandoc/Portfile]: Stack builds work using native `gcc`, but break using out-of-the-box MacPorts gcc.

I’m sure there’s a fix, but I never had time to track it down so just punted and used native `gcc` until someone can figure out the fix.

File: textproc/pandoc/Portfile (hunk 2 above)
Comment (essandess) [reply textproc/pandoc/Portfile]: I’ll go with consensus. I think of `configure` as the step that makes the Makefile, which isn’t being done by `setup`, but it is analogous.

File: textproc/pandoc/Portfile (hunk 1 above)
Comment (mojca) [reply textproc/pandoc/Portfile]: We don't want to build anything with the "real" GNU `gcc`, and in case the build was calling gnu gcc from MacPorts, that was a flaw that shouldn't have happened.

What you call "native `gcc`" (`/usr/bin/gcc`) is in fact clang on any modern macOS.

Can you please try using `${configure.compiler}` at the place where you used `/usr/bin/gcc` (or perhaps replace `system_gcc` with `configure.compiler` and maybe don't try to change the `PATH`)? In what way does that fail? (You should leave all explicit settings of the compiler as you have them now, just use `${configure.compiler}` instead of `/usr/bin/gcc`.)

Comment (cjones051073) [reply textproc/pandoc/Portfile]: Completely agree with @mojca. Please use `${configure.compiler}` and if there are issues they should be addressed. hardcoding `/usr/bin/gcc` is really not an acceptable solution.

Comment (essandess) [reply textproc/pandoc/Portfile]: Okay, I replaced with `${configure.compiler}`.

If this builds, I'll single-step through the other comments and make sure each delta produces a successful build.

I'm also wondering about the necessity of those `.env` statements too …

File: textproc/pandoc/Portfile (hunk 3 above)
Comment (essandess) [reply textproc/pandoc/Portfile]: Okay, but what am I supposed to say? I though that writing the `test` block did this, but no?

File: textproc/pandoc/Portfile (hunk 1 above)
Comment (essandess) [reply textproc/pandoc/Portfile]: `stack` needs the whole path to the compiler executable, not simply "`clang`".

How is this done within a Portfile?

See: https://paste.z0k.xyz/8b7e952d4573
```
DEBUG: system -W /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_textproc_pandoc/pandoc/work/pandoc-2.7.3: /opt/local/bin/stack setup --stack-root /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_textproc_pandoc/pandoc/work/.stack  --with-gcc clang --allow-different-user  && /opt/local/bin/stack build --stack-root /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_textproc_pandoc/pandoc/work/.stack  --with-gcc clang --allow-different-user  && /opt/local/bin/stack install --stack-root /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_textproc_pandoc/pandoc/work/.stack  --local-bin-path /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_textproc_pandoc/pandoc/work/pandoc-2.7.3/bin  --with-gcc clang --allow-different-user
option --with-gcc: InvalidAbsFile "clang"
```

Comment (cjones051073) [reply textproc/pandoc/Portfile]: try with `${configure.cc}` that gives a complete path.

File: textproc/pandoc/Portfile (hunk 3 above)
Comment (cjones051073) [reply textproc/pandoc/Portfile]: All the above does is define what running a test is supposed to do. It does not activate running it by default. For that just add

```
test.run yes
```

to the port file

File: textproc/pandoc/Portfile (hunk 4 above)
Comment (cjones051073) [reply textproc/pandoc/Portfile]: yes, that would be better than refining the test block en-mass. Between `test.cmd` `test.args` and `test.pre_args` it should be possible.

Comment (cjones051073) [reply textproc/pandoc/Portfile]: there is also `test.dir` if you need it..

File: textproc/pandoc/Portfile (hunk 1 above)
Comment (mojca) [reply textproc/pandoc/Portfile]: Please excuse me, @cjones051073 is right, it should be `configure.cc` instead. I would be surprised if that doesn't set the full path to the compiler.

Comment (cjones051073) [reply textproc/pandoc/Portfile]: It does.

File: textproc/pandoc/Portfile (hunk 2 above)
Comment (cjones051073) [reply textproc/pandoc/Portfile]: I agree with @mojca here. Configure phase is not just for making the make file, by any checks and configurations that are performed before the actual build starts. The first step here, setup, looks like it should be done in this phase. Likewise the final install step should be moved to the destroot phase.

File: textproc/pandoc/Portfile [hunk 5]
```diff
@@ -24,33 +25,43 @@ long_description    \
...
+depends_build-append  \
+                    port:stack
+
+use_configure       no
+
+# the PATH environment must provide /usr/bin/gcc, not ${prefix}/bin/gcc
+build.env \
+    PATH=/usr/bin:$env(PATH) \
+    PREFIX=${prefix} \
+    CC=${configure.cc}
```
Comment (mojca) [inline textproc/pandoc/Portfile]: Cool, now that `configure.cc` works, do you still need to override `build.env`? What happens if you simply remove `build.env`? (Try it locally first. I'm not behind my mac at the moment to be able to test.)

I would imagine that macports already sets all these three variables correctly.

File: textproc/pandoc/Portfile [hunk 6]
```diff
@@ -24,33 +25,43 @@ long_description    \
...
+
+build {
+    xinstall -m 0755 -d ${workpath}/.stack
```
Comment (mojca) [inline textproc/pandoc/Portfile]: You could create this folder in `post-extract`. It will then be easier to just run the three commands nearly out-of-the-box inside configure-build-destroot phase.

You would just override `build.cmd ${prefix}/bin/stack`, `build.target build` and provide whatever `build.args` might be needed.

File: textproc/pandoc/Portfile (hunk 1 above)
Comment (essandess) [reply textproc/pandoc/Portfile]: That worked. Next I’ll implement the test stuff so it catches any issues when I implement the rest of the changes.

Comment (mojca) [reply textproc/pandoc/Portfile]: > BTW. The normal way we handle situations like that is to use `compiler.blacklist`.

Sure, but only once you have [TheRightCompiler](https://trac.macports.org/wiki/UsingTheRightCompiler) working.

File: textproc/pandoc/Portfile (hunk 5 above)
Comment (mojca) [reply textproc/pandoc/Portfile]: Just to clarify:
* buildbot is only running the builds after the changes end up in repository (https://build.macports.org)
* for pull requests we use Travis and Azure
  * if Azure succeeds, you may ignore Travis as Travis too often fails for irrelevant reasons

File: textproc/pandoc/Portfile [hunk 7]
```diff
@@ -24,33 +25,36 @@ long_description    \
...
+depends_build-append  \
+                    port:stack
+
+set stack_root      ${workpath}/.stack
+
+post-extract {
+    xinstall -m 0755 -d ${stack_root}
+}
+
+configure.cmd       stack
```
Comment (mojca) [inline textproc/pandoc/Portfile]: I would use full path just in case (`${prefix}/bin/stack`).

Comment (mojca) [reply textproc/pandoc/Portfile]: I didn't want to suggest that this would be the reason for the failure. Just that it's in general a good practice to do it.

File: textproc/pandoc/Portfile (hunk 5 above)
Comment (mojca) [reply textproc/pandoc/Portfile]: It's highly likely that it already works if you say
```
sudo port -v install pandoc configure.compiler=macports-clang-8.0
```
for example. Or at least it should if everything was set up correctly.

Blacklisting gcc compiler if it doesn't work is another thing. (There's absolutely no need to fix anything in that respect at the moment.)

File: textproc/pandoc/Portfile (hunk 2 above)
Comment (mojca) [reply textproc/pandoc/Portfile]: > I had to hack a nonexistent `configure.target` with `configure.args`. Is there a better way?

If it doesn't exist, it might be a valid request to ask for official support of `configure.target` in the future. For now you probably just need to place it inside one of the 'arguments' of `configure` (`configure.cmd`, `configure.pre-args`, `configure.args`, ... whatever makes more sense), so I guess that it's ok what you have now.

**Comment by mojca** [comment]
I wanted to test, but my landline internet connection is broken, and over mobile it seems to be stuck after downloading 71 or 81 MB of data, with no further data transfer being made, and obviously no relevant logs that I would see anywhere. This is sadly quite a bit of a drawback of the current method (but still better than having no `pandoc` available, I guess).

**Comment by mojca** [comment]
@essandess: there is no need for a separate PR if you simply add the changes here.

**Comment by mojca** [comment]
I'm not even sure if the number of bytes transferred is the problem. It's simply that nothing happens for a very long period of time, so I'm not even sure if anything is being transferred at all, it seems to be stuck at `Downloading index`. How big is your `.stack` folder?

File: _resources/port1.0/group/haskell-stack-1.0.tcl [hunk 8]
```diff
@@ -0,0 +1,32 @@
+depends_build-append  \
```
Comment (cjones051073) [inline _resources/port1.0/group/haskell-stack-1.0.tcl]: Please add the first line shebang to this new file. Also, could you please add a brief description at the top of the port documenting its purpose and basic usage in ports ?

**Comment by mojca** [comment]
My question was more of: how much data needs to be transferred to be able to build pandoc? Or: how big is the `.stack` folder inside pandoc build before it gets purged?

File: _resources/port1.0/group/haskell-stack-1.0.tcl (hunk 8 above)
Comment (neverpanic) [reply _resources/port1.0/group/haskell-stack-1.0.tcl]: Done.

**Comment by neverpanic** [comment]
Unfortunately it seems this won't build (at least not without trace mode) if the `libiconv` port is installed. Building with trace mode also fails because we need an exception for `/var/db/mds/system` (which I plan on adding).

I'll merge this anyway as soon as the CI is done, but we should fix this in a follow-up.

**Comment by neverpanic** [comment]
The Travis CI is way behind again, so I'll merge this as-is now.

**Comment by mojca** [comment]
According to [buildbot](https://build.macports.org/builders/ports-10.13_x86_64-builder/builds/62891/steps/install-port/logs/statistics) it's just 4 GB to build `pandoc`.

**Comment by cjones051073** [comment]
I don't see any need to revert, thats always a drastic measure involving epochs etc.. The issue is just a build issue with `libiconv`. The binary tarballs are fine, and anyone building fro  source can just temporarily uninstall `libiconv` to side-step the problem.

**Comment by cjones051073** [comment]
> @neverpanic @mojca @cjones051073 @mf2k
> 
> `LIBRARY_PATH=/opt/local/lib` is the offending variable.
> 
> How do I write a `configure.env` that prevents this variable from being set? Or overwrites it?

what library exactly is being picked up from `/opt/local/lib` that is causing the problems ?

**Comment by mojca** [comment]
@pmetzger: I'm pretty sure this will become broken if you revert. Updating from binary should work just fine.

**Comment by cjones051073** [comment]
> > I don't see any need to revert, thats always a drastic measure involving epochs etc.. The issue is just a build issue with `libiconv`. The binary tarballs are fine, and anyone building fro source can just temporarily uninstall `libiconv` to side-step the problem.
> 
> @cjones051073 is right, @pmetzger. The new `pandoc` binary upgrade works perfectly.
> 
> The only thing that’s busted is building from source if you have MacPorts `libiconv`. installed. Turning off the default MacPorts `LIBRARY_PATH='/opt/local/lib'` will fix this.

removing LIBRARY_PATH='/opt/local/lib' is overly drastic, in my opinion.

If the issue is just a build conflict against `libiconv` then instead use the `conflicts_build` PG to declare this conflict. Doesn't fix it, but gives the user building from source a clear message.

**Comment by cjones051073** [comment]
> @mf2k
> 
> > Have you tried this?
> > `configure.env-delete LIBRARY_PATH='/opt/local/lib'`
> 
> Thanks, tried it, but `LIBRARY_PATH` is still getting set.
> 
> Side comment: in all the mucking around I’ve done with Portfile `.env` settings, I’ve never observed them to actually change the environment variables.

what phase are you looking in ? If you want to change the ENV during the build phase, you need to munge around with `build.env` ;)

**Comment by neverpanic** [comment]
A quick grep of the macports-base source reveals where `LIBRARY_PATH` gets set: https://github.com/macports/macports-base/blob/master/src/port1.0/portutil.tcl#L425-L427

Consequently, you'll have to clear `compiler.library_path` for this to disappear.

Btw, the problem with libiconv only shows up if the headers do not match the linked libraries – that's because the macports copy of libiconv (which is exactly the upstream copy) transparently redefines any use of `iconv*` symbols to `libiconv*` (which is what the macports-installed library provides). Apple's headers in `/usr/include` don't do this. This suggests that the headers were used from `/usr/include`, but linking attempts to use `/opt/local/lib`.

Of course this may be an artifact of downloading pre-built binaries which have already consumed the headers from `/usr/include`.

FWIW, enabling trace mode with https://github.com/macports/macports-base/pull/144 merged also works around this. Considering I expected `pandoc` to produce a binary-distributable archive I elected to merge this PR regardless of this problem.
